src/herder/Upgrades.cpp

```diff
@@ -318,4 +329,348 @@ Upgrades::timeForUpgrade(uint64_t time) const
+            {
+                // This is equivalent to adjustOffer, as the selling liabilities
+                // of an offer should equal the adjusted quantity.
+                // TODO(jonjove): Test the above claim
+                offer.amount = offerFrame->getSellingLiabilities();
+                offerFrame->storeChange(ld, db);
+
+                if (offer.buying.type() == ASSET_TYPE_NATIVE ||
+                    !(offer.sellerID == getIssuer(offer.buying)))
+                {
+                    if (!stellar::addBalance(
+                            liabilities[offer.buying].buying,
+                            offerFrame->getBuyingLiabilities()))
+                    {
+                        throw std::runtime_error("could not add buying "
+                                                 "liabilities");
+                    }
+                }
+                if (offer.selling.type() == ASSET_TYPE_NATIVE ||
+                    !(offer.sellerID == getIssuer(offer.selling)))
+                {
+                    if (!stellar::addBalance(
+                            liabilities[offer.selling].selling,
+                            offerFrame->getSellingLiabilities()))
+                    {
+                        throw std::runtime_error("could not add selling "
+                                                 "liabilities");
+                    }
+                }
+            }
+        }
+
+        for (auto const& assetLiabilities : liabilities)
+        {
+            Asset const& asset = assetLiabilities.first;
+            Liabilities const& liab = assetLiabilities.second;
+            if (asset.type() == ASSET_TYPE_NATIVE)
+            {
+                int64_t deltaSelling =
+                    liab.selling -
+                    accountFrame->getSellingLiabilities(ledgerManager);
+                int64_t deltaBuying =
+                    liab.buying -
+                    accountFrame->getBuyingLiabilities(ledgerManager);
+                if (!accountFrame->addSellingLiabilities(deltaSelling,
+                                                         ledgerManager))
+                {
+                    throw std::runtime_error("invalid selling liabilities "
+                                             "during upgrade");
+                }
+                if (!accountFrame->addBuyingLiabilities(deltaBuying,
+                                                        ledgerManager))
+                {
+                    throw std::runtime_error("invalid buying liabilities "
+                                             "during upgrade");
+                }
+            }
+            else
+            {
+                auto trustFrame = TrustFrame::loadTrustLine(accountOffers.first,
+                                                            asset, db, &ld);
+                int64_t deltaSelling =
+                    liab.selling -
+                    trustFrame->getSellingLiabilities(ledgerManager);
```

Comment:
This errors when there are orphaned offers:
```
Exception during upgrade: trustline does not exist [LedgerManagerImpl.cpp:880]
```

Repro:
1. Run standalone testnet (testnet passphrase).
2. Submit:
  ```
  AAAAAGL8HQvQkbK2HA3WVjRrKmjX00fG8sLI7m0ERwJW/AX3AAABkAAAAAAAAAABAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAADJNEgW9b6gnIKGtYeSq1SiCyMJLujoFBoDtnljGf65oAAAAXSHboAAAAAAEAAAAADJNEgW9b6gnIKGtYeSq1SiCyMJLujoFBoDtnljGf65oAAAAGAAAAAVVTRAAAAAAAYvwdC9CRsrYcDdZWNGsqaNfTR8bywsjubQRHAlb8BfcAAAAAO5rKAAAAAAEAAAAADJNEgW9b6gnIKGtYeSq1SiCyMJLujoFBoDtnljGf65oAAAADAAAAAAAAAAFVU0QAAAAAAGL8HQvQkbK2HA3WVjRrKmjX00fG8sLI7m0ERwJW/AX3AAAAAACYloAAAAABAAAAAQAAAAAAAAAAAAAAAQAAAAAMk0SBb1vqCcgoa1h5KrVKILIwku6OgUGgO2eWMZ/rmgAAAAYAAAABVVNEAAAAAABi/B0L0JGythwN1lY0aypo19NHxvLCyO5tBEcCVvwF9wAAAAAAAAAAAAAAAAAAAAJW/AX3AAAAQK6ngzT5yLjXBKPibogU0NBm9M+W+ROt2JcPiIe1mf9eGNGaWhj0DYyG3GkTzrN4Tzg+LpXhiEO4cu6gylK6bQkxn+uaAAAAQAADEbTeCGHZ3RQVmUOX2pM8J+lFr5cq1PjOQZtoxSy1jYTnrvoRHoc7z9eSj9tv8aU/9ufEYBUYawc/l6hLMAI=
  ```
3. Upgrade to V10.